File: Modules/EBanking/Transactions/TransactionsForm.cs
using cyberbanking.EBanking.Accounts;
using Serenity.ComponentModel;
using System;
using System.ComponentModel;

namespace cyberbanking.EBanking.Forms;

[FormScript("EBanking.Transactions")]
[BasedOnRow(typeof(TransactionsRow), CheckNames = true)]
[ReadOnly(true)]
public class TransactionsForm
{
    public decimal Amount { get; set; }

    public int TransactionType { get; set; }

    [Hidden]
    public DateTime TransactionDate { get; set; }

    [TextAreaEditor(Rows = 5)]
    public string Description { get; set; }

    [HideOnInsert, HideOnUpdate]
    public int SenderAccountId { get; set; }

    public AccountType SenderAccountType { get; set; }

    public int ReceiverAccountId { get; set; }
    public AccountType ReceiverAccountType { get; set; }
}

File: Modules/ServerTypes/EBanking/TransactionsForm.ts
import { DecimalEditor, EnumEditor, DateEditor, TextAreaEditor, LookupEditor, PrefixedContext } from "@serenity-is/corelib";
import { TransactionType } from "./Transactions.TransactionType";
import { AccountType } from "./Accounts.AccountType";
import { initFormType } from "@serenity-is/corelib/q";

export interface TransactionsForm {
    Amount: DecimalEditor;
    TransactionType: EnumEditor;
    TransactionDate: DateEditor;
    Description: TextAreaEditor;
    SenderAccountId: LookupEditor;
    SenderAccountType: EnumEditor;
    ReceiverAccountId: LookupEditor;
    ReceiverAccountType: EnumEditor;
}

export class TransactionsForm extends PrefixedContext {
    static formKey = 'EBanking.Transactions';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!TransactionsForm.init)  {
            TransactionsForm.init = true;

            var w0 = DecimalEditor;
            var w1 = EnumEditor;
            var w2 = DateEditor;
            var w3 = TextAreaEditor;
            var w4 = LookupEditor;

            initFormType(TransactionsForm, [
                'Amount', w0,
                'TransactionType', w1,
                'TransactionDate', w2,
                'Description', w3,
                'SenderAccountId', w4,
                'SenderAccountType', w1,
                'ReceiverAccountId', w4,
                'ReceiverAccountType', w1
            ]);
        }
    }
}

[TransactionType, AccountType]; // referenced types